pkg/runners/zero/zero.go

```diff
@@ -240,6 +268,98 @@ func (runner *ZeroRunner) RunFor(steps uint64) error {
 	return nil
 }
 
+// EndRun is responsible for running the additional steps after the program was executed,
+// until the checkUsedCells doesn't return any error.
+// Since this vm always finishes the run of the program at the number of steps that is a power of two in the proof mode,
+// there is no need to run additional steps before the loop.
+func (runner *ZeroRunner) EndRun() {
+	if runner.proofmode {
+		for runner.checkUsedCells() != nil {
+			pow2Steps := utils.NextPowerOfTwo(runner.vm.Step + 1)
+			if err := runner.RunFor(pow2Steps); err != nil {
+				panic(err)
+			}
+		}
+	}
+}
+
+// checkUsedCells returns error if not enough steps were made to allocate required number of cells for builtins
+// or there are not enough trace cells to fill the entire range check range
+func (runner *ZeroRunner) checkUsedCells() error {
+	for _, bRunner := range runner.Layout.Builtins {
+		builtinSegment, ok := runner.vm.Memory.FindSegmentWithBuiltin(bRunner.Runner.String())
+		if ok {
+			_, err := bRunner.Runner.GetAllocatedSize(builtinSegment.Len(), runner.steps())
```

Comment:
yes exactly, I just check for an error. If such error appears, we need to run for more steps. Since each builtin has assigned ratio (estimated value of how many instances will appear per n number of steps), then if we have more builtins then the estimated value, we need to run for more steps)